# Линейная классификация и логистическая регрессия (часть 2)

Коптев Данил

**Подключим библиотеки**

In [34]:
from sklearn import datasets
import numpy as np
from sklearn.linear_model import SGDClassifier
from sklearn.model_selection import ParameterGrid, train_test_split, GridSearchCV, RandomizedSearchCV

**Задание 4**

1. Изучим  самостоятельно  инструментарий  подбора  гиперпараметров  модели, реализованный в библиотеке sklearn



2. Используя инструментарий модуля datasets, импортировать встроенный набор данных iris(часто используется авторами учебников и учебных курсов по машинному обучению в качестве учебного набора данных; содержит информацию о 150 объектах –цветах ириса, которые на основании значений четырех признаков необходимо отнести к одному из трех сортов).О наборе данных. Описание набора в sklearn.Формат команды: <имя набора данных> = datasets.load_iris()Документация метода load_iris().

In [35]:
iris = datasets.load_iris()
iris.keys()

dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names', 'filename', 'data_module'])

3. В результате выполнения п. 2 должен быть создан объект, по структуре подобный словарю. Чтобы убедиться в этом и получить представление о структуре набора данных, вывести наименования ключей набора. 

4. Вывести значение ключа DESCR(содержит описание набора данных) и изучить это описание.

In [36]:
print(iris['DESCR'])

.. _iris_dataset:

Iris plants dataset
--------------------

**Data Set Characteristics:**

    :Number of Instances: 150 (50 in each of three classes)
    :Number of Attributes: 4 numeric, predictive attributes and the class
    :Attribute Information:
        - sepal length in cm
        - sepal width in cm
        - petal length in cm
        - petal width in cm
        - class:
                - Iris-Setosa
                - Iris-Versicolour
                - Iris-Virginica
                
    :Summary Statistics:

    ============== ==== ==== ======= ===== ====================
                    Min  Max   Mean    SD   Class Correlation
    ============== ==== ==== ======= ===== ====================
    sepal length:   4.3  7.9   5.84   0.83    0.7826
    sepal width:    2.0  4.4   3.05   0.43   -0.4194
    petal length:   1.0  6.9   3.76   1.76    0.9490  (high!)
    petal width:    0.1  2.5   1.20   0.76    0.9565  (high!)
    ============== ==== ==== ======= ===== ===========

5. Изучить значения остальных ключей и определить ключи, в которых хранятся значения признаков и метки классов. Далее вся работа –только со значениями этих ключей.

In [37]:
data = iris.data
data

array([[5.1, 3.5, 1.4, 0.2],
       [4.9, 3. , 1.4, 0.2],
       [4.7, 3.2, 1.3, 0.2],
       [4.6, 3.1, 1.5, 0.2],
       [5. , 3.6, 1.4, 0.2],
       [5.4, 3.9, 1.7, 0.4],
       [4.6, 3.4, 1.4, 0.3],
       [5. , 3.4, 1.5, 0.2],
       [4.4, 2.9, 1.4, 0.2],
       [4.9, 3.1, 1.5, 0.1],
       [5.4, 3.7, 1.5, 0.2],
       [4.8, 3.4, 1.6, 0.2],
       [4.8, 3. , 1.4, 0.1],
       [4.3, 3. , 1.1, 0.1],
       [5.8, 4. , 1.2, 0.2],
       [5.7, 4.4, 1.5, 0.4],
       [5.4, 3.9, 1.3, 0.4],
       [5.1, 3.5, 1.4, 0.3],
       [5.7, 3.8, 1.7, 0.3],
       [5.1, 3.8, 1.5, 0.3],
       [5.4, 3.4, 1.7, 0.2],
       [5.1, 3.7, 1.5, 0.4],
       [4.6, 3.6, 1. , 0.2],
       [5.1, 3.3, 1.7, 0.5],
       [4.8, 3.4, 1.9, 0.2],
       [5. , 3. , 1.6, 0.2],
       [5. , 3.4, 1.6, 0.4],
       [5.2, 3.5, 1.5, 0.2],
       [5.2, 3.4, 1.4, 0.2],
       [4.7, 3.2, 1.6, 0.2],
       [4.8, 3.1, 1.6, 0.2],
       [5.4, 3.4, 1.5, 0.4],
       [5.2, 4.1, 1.5, 0.1],
       [5.5, 4.2, 1.4, 0.2],
       [4.9, 3

In [38]:
target = iris.target
target

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

6. Выполнить  разбиение  набора  данных  (значения  признаков  и  метки  классов)  на обучающую и тестовую выборки в соотношении 70/30.

In [39]:
x_train, x_test, y_train, y_test = train_test_split(data, target, test_size=0.3, random_state=25)

7. Создать модель линейной классификации SGDClassifierи вывести список доступных параметров модели.

In [40]:
SGDC = SGDClassifier()
SGDC.get_params()

{'alpha': 0.0001,
 'average': False,
 'class_weight': None,
 'early_stopping': False,
 'epsilon': 0.1,
 'eta0': 0.0,
 'fit_intercept': True,
 'l1_ratio': 0.15,
 'learning_rate': 'optimal',
 'loss': 'hinge',
 'max_iter': 1000,
 'n_iter_no_change': 5,
 'n_jobs': None,
 'penalty': 'l2',
 'power_t': 0.5,
 'random_state': None,
 'shuffle': True,
 'tol': 0.001,
 'validation_fraction': 0.1,
 'verbose': 0,
 'warm_start': False}

8. Создать  сетку  параметров,  включающую  4  вида  функции  потерь,  два  типа регуляризаторов, 5 значений коэффициента регуляризации от 0.0001 до 0.001, и число итераций от 5 до 10 с шагом 1.

9. Создать объект GridSearchCV, передать ему созданный ранее классификатор и сетку параметров и обучить его. В качестве метрики использовать accuracy. Предусмотреть вывод  времени,  затраченного  на  перебор  по  сетке  (можно  использовать команду%%time).

In [41]:
%%time

gridSearch = GridSearchCV(
    SGDC, 
    {
    'loss': ['log_loss', 'squared_hinge', 'hinge', 'modified_huber'],
    'penalty': ['l1', 'l2'],
    'alpha': np.linspace(0.0001, 0.001, 5),
    'n_iter_no_change': np.arange(5,11)
    }, 
    cv=5, 
    scoring='accuracy')
gridSearch.fit(x_train, y_train)

CPU times: user 4.41 s, sys: 42.8 ms, total: 4.45 s
Wall time: 4.46 s


GridSearchCV(cv=5, estimator=SGDClassifier(),
             param_grid={'alpha': array([0.0001  , 0.000325, 0.00055 , 0.000775, 0.001   ]),
                         'loss': ['log_loss', 'squared_hinge', 'hinge',
                                  'modified_huber'],
                         'n_iter_no_change': array([ 5,  6,  7,  8,  9, 10]),
                         'penalty': ['l1', 'l2']},
             scoring='accuracy')

10. Вывести  лучший  классификатор,  лучший  набор  параметров  и  оценку  лучшего классификатора в соответствии с заданной метрикой. Получить оценки алгоритма по последним 10 наборам параметров.

In [42]:
print("Лучший классификатор: ", gridSearch.best_estimator_, "\n")
print("Лучший набор параметров: ", gridSearch.best_params_, "\n")
print("Оценка лучшего классификатора: ", gridSearch.best_score_, "\n")
print("Оценки алгоритма по последним 10 наборам параметров",  gridSearch.cv_results_['mean_test_score'][:10])

Лучший классификатор:  SGDClassifier(alpha=0.0007750000000000001, n_iter_no_change=9, penalty='l1') 

Лучший набор параметров:  {'alpha': 0.0007750000000000001, 'loss': 'hinge', 'n_iter_no_change': 9, 'penalty': 'l1'} 

Оценка лучшего классификатора:  0.980952380952381 

Оценки алгоритма по последним 10 наборам параметров [0.80952381 0.79047619 0.8952381  0.76190476 0.9047619  0.83809524
 0.85714286 0.83809524 0.83809524 0.85714286]


11. Проанализировать полученные результаты, сделать выводы.

12. Организовать случайный поиск по сетке. Вывести время, затраченное на случайный перебор.

**Анализ:** самый лучший результат показал классификатор SGDClassifier с точность 0.980952380952381, что является достаточном высоким показателем. Для его достижения был использован регуляризаторо $L_1$ коэфицентом регуляризации 0.000775, с показатель остановки  9, а функция потерь была использована hinge. Общее время реализации алгоритма 4.46 s.

Говоря о оценки алгоритма по последним 10 наборам параметров, можно сказать, что значения имеют довольно большой разборс, что говорит о большом объеме неудовлетворительных комбинациях, то есть для достижения наилучшего классификатора алгоритму необходимы испродобавать большое колличество паратемров, но наилучший есть риск так и не найти, поэтому есть необходимость использовать более обширный перечень параметров 

13. Вывести те же показатели, что и в п. 10. Сопоставить полученные результаты и время на перебор с результатами, полученными при выполнении п. 10.

In [43]:
%%time

randomizedSearch = RandomizedSearchCV(
    SGDC, 
    {
    'loss': ['hinge', 'log_loss', 'modified_huber', 'squared_hinge'],
    'penalty': ['l1', 'l2'],
    'alpha': np.linspace(0.0001, 0.001, 5),
    'n_iter_no_change': np.arange(5,11)
    }, 
    n_iter=50, 
    cv=5, 
    scoring='accuracy', 
    random_state=25)
randomizedSearch.fit(x_train, y_train)

CPU times: user 931 ms, sys: 9.56 ms, total: 940 ms
Wall time: 937 ms


RandomizedSearchCV(cv=5, estimator=SGDClassifier(), n_iter=50,
                   param_distributions={'alpha': array([0.0001  , 0.000325, 0.00055 , 0.000775, 0.001   ]),
                                        'loss': ['hinge', 'log_loss',
                                                 'modified_huber',
                                                 'squared_hinge'],
                                        'n_iter_no_change': array([ 5,  6,  7,  8,  9, 10]),
                                        'penalty': ['l1', 'l2']},
                   random_state=25, scoring='accuracy')

In [44]:
print("Лучший классификатор: ", randomizedSearch.best_estimator_, "\n")
print("Лучший набор параметров: ", randomizedSearch.best_params_, "\n")
print("Оценка лучшего классификатора: ", randomizedSearch.best_score_, "\n")
print("Оценки алгоритма по последним 10 наборам параметров", randomizedSearch.cv_results_['mean_test_score'][:10])

Лучший классификатор:  SGDClassifier(alpha=0.001, loss='log_loss', n_iter_no_change=8, penalty='l1') 

Лучший набор параметров:  {'penalty': 'l1', 'n_iter_no_change': 8, 'loss': 'log_loss', 'alpha': 0.001} 

Оценка лучшего классификатора:  0.9714285714285715 

Оценки алгоритма по последним 10 наборам параметров [0.83809524 0.78095238 0.93333333 0.9047619  0.93333333 0.92380952
 0.91428571 0.8952381  0.96190476 0.93333333]


14. Сделать выводы, сформулировать рекомендации.

**Анализ:** самый лучший результат показал классификатор SGDClassifier с точность $0.9714285714285715$ (с ипользованием GridSearchCV: 0.980952380952381), что является достаточном высоким показателем. Для его достижения был использован регуляризаторо $L_1$ (с ипользованием GridSearchCV: $L_1$)  коэфицентом регуляризации $0.001$ (с ипользованием GridSearchCV: 0.000775), с показатель остановки  $8$ (с ипользованием GridSearchCV: 9), а функция потерь была использована **log_loss** (с ипользованием GridSearchCV: hinge). 

Говоря о оценки алгоритма по последним 10 наборам параметров, можно сказать, что значения имеют довольно большой разборс, что говорит о больном объемне неудовлетворительных комбинациях, а так же можно заметить, что разброс значений значительно увеличился по сравнению с GridSearchCV, но это объясняет о случайном происхождении параметров.

Обращая внимания на время выполнения, то мы наглядным образом убедились в том, что GridSearchCV работает дольше, чем RandomizedSearchCV, но в данном примере эффиктивность пусть и незначительно, но упала.

**Рекомендация**: затраченное время на испольнения алгоритма GridSearchCV связана с колличеством данных, поэтому можно утверждать, что использование GridSearchCV может быть с использованием небольшим объемом данных, иначе возрастает риск увелечение времени отработки алгоритма на довольно продолжительное время, но можно так же утвержать, что точность данных моделей будет более выгодна, чем при использование RandomizedSearchCV. Так же необходимо использовать более объемные комбинации параметров для нахождения наиболее предпочтительного решения в том или ином наборе данных внезависимости от того выбираете ли вы GridSearchCV или RandomizedSearchCV.